src/main/java/org/assertj/core/api/AbstractAssert.java

```diff
@@ -358,105 +360,112 @@ private void printDescriptionText() {
   /** {@inheritDoc} */
   @Override
   public SELF isEqualTo(Object expected) {
-    objects.assertEqual(info, actual, expected);
+    //return doAssert(innerExpected -> objects.assertEqual(info, actual, innerExpected), expected);
+    return doAssert(() -> objects.assertEqual(info, actual, expected));
```

Comment:
another option would be:
```java
 public SELF isEqualTo(Object expected) {
    if (ignoreWhen != null && ignoreWhen.test(actual)) return myself;
    return objects.assertEqual(info, actual, expected);
}
```
This is not much more work that  wrapping the call in `doAssert` and it has low performance impact for tests not using `ignoreWhen(Predicate<ACTUAL> predicate)` since a simple null check is performed vs a predicate call.

`ignoreWhen != null && ignoreWhen.test(actual)` can be refactored into a `ignoreAssertion(actual)` method:
```java
 public SELF isEqualTo(Object expected) {
    if (ignoreAssertion(actual)) return myself;
    return objects.assertEqual(info, actual, expected);
}
```

Reply:
I think we should run a perf tests before and after to see the real impact of any approach, otherwise it's just gesstimating the impact.

Reply:
You are right. I'll try to do some perf tests as well. Do you by any chance have some repo where you do some benchmarks already?

Reply:
nope :( we have [assertj-examples](https://github.com/assertj/assertj-examples/tree/main/assertions-examples/src/test/java/org/assertj/examples) or maybe it's time to create one.

Reply:
I'll try to see if I can come up with a base example first for `isEqualTo` and `isNotEqualTo` we can then go from there :)

Reply:
Have a look at https://github.com/filiphr/assertj-benchmark. It is extremely work in progress, but I think it could be beneficial. I am more than willing to transfer the repo to the assertj organization.

Reply:
So first benchmark results are in

![image](https://user-images.githubusercontent.com/6012094/117566185-7a28de80-b0b5-11eb-84d8-ba2a9e827e36.png)

Left is with the changes in this PR and right is with 3.19.0. So there is some performance impact. Although I find it interesting the it is faster for `isNotEqualTo`. Although there is a difference for how the `isEqualTo` and `isNotEqualTo` are done.

Reply:
thanks, I'm happy to move your repo to the assertj organization and give you commit rights to.

Reply:
There is also a slight difference between `isEqualTo` and `isNotEqualTo`. I will try to align them and run the benchmark again.

For the repo, let me know how you want to do it. In order for me to move it I need to be part of the assertj organization. Another option would be you to create a repo and I can push to that one. I will leave this up to you

Reply:
I got a bit sidetracked with the benchmarks 😄. I realized that the drop in the `isEqualTo` is not 100% due to my changes, there is a drop due to https://github.com/assertj/assertj-core/commit/f2acb419eade8ab6704b62e8bf9d86c0104a643d. I also changed the benchmark a bit in order not to use a blackhole. And the latest results are the following:

![image](https://user-images.githubusercontent.com/6012094/118350242-e1331100-b555-11eb-8fbc-d7468a016399.png)


* resuls.png -> 3.19.0
* results_main.png -> The code from the main branch
* results_conditional.png -> The code from this PR rebased on the main branch

Looking at the latest numbers. The drop is extremely significant (50%). Therefore, I will try to see if we can do better.